File: sem6/seminar/classes.zig
const Animal = union(enum) {
    Dog: void,
    Cat: void,

    pub fn speak(self: Animal)const u8 {
        return switch (self) { 
           .Dog => "Woof",
           .Cat => "Meow",
        };
    }
};

// Values can be contiguous in memory (no hidden heap boxing)
const zoo = [_]Animal{.{.Dog = {} } };

for (zoo) |animal| {
    _ = animal.speak();
}

File: sem6/seminar/classes.rs
trait Animal { fn speak(&self) -> &str; }
struct Dog;
impl Animal for Dog { fn speak(&self) -> &str { "Woof" } }

// Requires Box (heap allocation) for dynamic sizes
let zoo: Vec<Box<dyn Animal>> = vec!;

for animal in &zoo {
    animal.speak();
}
